File: src/murmur/cadence.py
"""Talk-vs-music scheduling policies (spec 03-02 §2.3).

The Director consults a ``CadencePolicy`` at each segment boundary and never
knows which mode is behind the seam. ``EveryNCadence`` (default) and
``RandomCadence`` are pure local policy — 0 tokens (master §7 pillar 1).
``BrainCadence`` is the opt-in exception: a one-shot cheap-model judgment that
hard-falls-back to a local policy on any failure, timeout, or nonsense answer,
so the radio never stalls on the network.
"""

from __future__ import annotations

import asyncio
import random
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any, Literal, Protocol, TypedDict, runtime_checkable

from .harness import BrainTool, Harness
from .prompts import CADENCE_INSTRUCTION, CADENCE_STATE_HEADER

TALK = "talk"
MUSIC = "music"

_DEFAULT_EVERY_N = 2
_DEFAULT_TIMEOUT_S = 8.0


@dataclass(frozen=True)
class CadenceState:
    """Local signals a policy may consult; extended as later specs add
    sources (activity/pacing in 07, ledger in 05). ``situation`` is a rendered
    text block (recent turns etc.) that only ``BrainCadence`` reads."""

    talks_since_music: int
    situation: str = ""


@runtime_checkable
class CadencePolicy(Protocol):
    """The scheduling seam the Director consumes (spec 03-02 §2.3): it never
    knows which mode is behind it. Tests inject scripted fakes."""

    async def next_kind(self, state: CadenceState) -> str: ...


class EveryNCadence:
    """Default: a song after every N talk segments. Deterministic, 0 tokens."""

    def __init__(self, n: int = _DEFAULT_EVERY_N) -> None:
        self._n = max(1, n)

    async def next_kind(self, state: CadenceState) -> str:
        return MUSIC if state.talks_since_music >= self._n else TALK


class RandomCadence:
    """Probability ``p`` per boundary, guarded: never before ``min_gap`` talks,
    always by ``max_gap`` (no wall-to-wall music, no endless talk). The RNG is
    injectable so tests are deterministic."""

    def __init__(
        self,
        p: float = 0.35,
        *,
        min_gap: int = 1,
        max_gap: int = 6,
        rng: random.Random | None = None,
    ) -> None:
        self._p = p
        self._min_gap = max(0, min_gap)
        self._max_gap = max(self._min_gap, max_gap)
        self._rng = rng if rng is not None else random.Random()

    async def next_kind(self, state: CadenceState) -> str:
        talks = state.talks_since_music
        if talks < self._min_gap:
            return TALK
        if talks >= self._max_gap:
            return MUSIC
        return MUSIC if self._rng.random() < self._p else TALK


class _ChooseResult(TypedDict):
    ok: Literal[True]
    kind: str


class _ChooseSegmentTool:
    """Terminal tool: the model commits to talk or music (spec 03-01 §2.1
    termination rule — structured output via the terminal tool's schema)."""

    name = "choose_segment"
    description = "Commit to the next segment kind. Call exactly once."
    input_schema: dict[str, Any] = {
        "type": "object",
        "properties": {
            "kind": {
                "type": "string",
                "enum": [TALK, MUSIC],
                "description": "what plays next",
            }
        },
        "required": ["kind"],
    }
    terminal = True

    async def run(self, args: Mapping[str, object]) -> _ChooseResult:
        raw = args.get("kind")
        return _ChooseResult(ok=True, kind=raw if isinstance(raw, str) else "")


class BrainCadence:
    """Opt-in: ask a cheap model to pace by feel (the sanctioned master §7
    pillar-1 exception). Any failure — exception, timeout, no terminal call,
    invalid kind — falls back to a local policy; the stream never stalls."""

    def __init__(
        self,
        brain: Harness,
        *,
        model: str,
        fallback: EveryNCadence | RandomCadence | None = None,
        timeout_s: float = _DEFAULT_TIMEOUT_S,
    ) -> None:
        self._brain = brain
        self._model = model
        self._fallback = fallback if fallback is not None else EveryNCadence()
        self._timeout_s = timeout_s
        self._tools: list[BrainTool] = [_ChooseSegmentTool()]

    async def next_kind(self, state: CadenceState) -> str:
        prompt = (
            f"{CADENCE_INSTRUCTION}\n"
            f"{CADENCE_STATE_HEADER}"
            f"- talk segments since the last song: {state.talks_since_music}\n"
            f"{state.situation}"
        )
        try:
            result = await asyncio.wait_for(
                self._brain.run_task(
                    "",
                    prompt,
                    tools=self._tools,
                    model=self._model,
                    max_turns=2,
                ),
                self._timeout_s,
            )
        except Exception:
            return await self._fallback.next_kind(state)
        kind = result.get("kind") if result else None
        if kind in (TALK, MUSIC):
            return str(kind)
        return await self._fallback.next_kind(state)


def build_cadence(
    mode: str,
    *,
    every_n: int,
    brain: Harness | None = None,
    model: str = "",
) -> CadencePolicy:
    """Construct the configured cadence mode (spec 03-02 §2.3)."""
    if mode == "every_n":
        return EveryNCadence(every_n)
    if mode == "random":
        return RandomCadence()
    if mode == "brain":
        if brain is None:
            raise ValueError("brain cadence requires a Harness-capable brain")
        return BrainCadence(brain, model=model, fallback=EveryNCadence(every_n))
    raise ValueError(
        f"unknown cadence_mode {mode!r}; expected 'every_n', 'random', or 'brain'"
    )

File: src/cadence.ts
// Talk-vs-music scheduling (spec 03-02 §2.3). The Director consults this seam at
// each segment boundary and never knows which mode is behind it.
//
// EveryN and Random are pure local policy — 0 tokens (master §7 pillar 1).
// BrainCadence is the opt-in exception: one cheap-model judgment per boundary,
// which hard-falls-back to a local policy on any failure, timeout, or nonsense
// answer, so the program never stalls waiting on the network.

import { tool } from '@anthropic-ai/claude-agent-sdk'
import { z } from 'zod'

import type { Activity } from './activity.ts'
import type { Harness } from './contracts.ts'
import { CADENCE_INSTRUCTION, CADENCE_STATE_HEADER } from './prompts.ts'

export type SegmentKind = 'talk' | 'music'

// Local signals a policy may consult. `situation` is rendered text only
// BrainCadence reads; `activity` is the spec-07 presence signal PacingCadence
// gates on.
export type CadenceState = {
  readonly talksSinceMusic: number
  readonly situation?: string
  readonly activity?: Activity
}

export interface CadencePolicy {
  nextKind(state: CadenceState): Promise<SegmentKind>
}

export class EveryNCadence implements CadencePolicy {
  private n: number

  constructor(n = 2) {
    this.n = Math.max(1, n)
  }

  async nextKind(state: CadenceState): Promise<SegmentKind> {
    return state.talksSinceMusic >= this.n ? 'music' : 'talk'
  }
}

export type RandomCadenceOptions = {
  p?: number
  minGap?: number
  maxGap?: number
  // Injected so tests are deterministic.
  random?: () => number
}

// Probability p per boundary, guarded: never before minGap talks, always by
// maxGap — no wall-to-wall music, no endless talk.
export class RandomCadence implements CadencePolicy {
  private p: number
  private minGap: number
  private maxGap: number
  private random: () => number

  constructor({ p = 0.35, minGap = 1, maxGap = 6, random = Math.random }: RandomCadenceOptions = {}) {
    this.p = p
    this.minGap = Math.max(0, minGap)
    this.maxGap = Math.max(this.minGap, maxGap)
    this.random = random
  }

  async nextKind(state: CadenceState): Promise<SegmentKind> {
    if (state.talksSinceMusic < this.minGap) return 'talk'
    if (state.talksSinceMusic >= this.maxGap) return 'music'
    return this.random() < this.p ? 'music' : 'talk'
  }
}

// Presence policy composes with whatever cadence the user chose (spec 07 §2.5)
// instead of replacing it: an away room gets music/bed, everything else is the
// configured policy's call. The short-circuit lands BEFORE the delegation, so
// an away room also skips the opt-in brain cadence — gating saves that token
// too (§3.3).
export class PacingCadence implements CadencePolicy {
  private inner: CadencePolicy

  constructor(inner: CadencePolicy) {
    this.inner = inner
  }

  async nextKind(state: CadenceState): Promise<SegmentKind> {
    if (state.activity === 'away') return 'music'
    return this.inner.nextKind(state)
  }
}

const DEFAULT_TIMEOUT_MS = 8_000

export type BrainCadenceOptions = {
  brain: Harness
  model: string
  fallback?: CadencePolicy
  timeoutMs?: number
}

export class BrainCadence implements CadencePolicy {
  private opts: BrainCadenceOptions
  private fallback: CadencePolicy

  constructor(opts: BrainCadenceOptions) {
    this.opts = opts
    this.fallback = opts.fallback ?? new EveryNCadence()
  }

  async nextKind(state: CadenceState): Promise<SegmentKind> {
    const prompt =
      `${CADENCE_INSTRUCTION}\n${CADENCE_STATE_HEADER}` +
      `- talk segments since the last song: ${state.talksSinceMusic}\n${state.situation ?? ''}`
    try {
      // A hung model must not hold the boundary: race the judgment against the
      // deadline. The abandoned task keeps running (promises cannot be
      // cancelled) but nothing waits on it.
      const kind = await Promise.race([
        this.opts.brain.runTask<SegmentKind>({
          systemPrompt: '',
          prompt,
          model: this.opts.model,
          maxTurns: 2,
          tools: (finish) => [chooseSegmentTool(finish)],
        }),
        new Promise<null>((resolve) =>
          setTimeout(() => resolve(null), this.opts.timeoutMs ?? DEFAULT_TIMEOUT_MS).unref(),
        ),
      ])
      if (kind !== null) return kind
    } catch {
      // fall through to local policy
    }
    return this.fallback.nextKind(state)
  }
}

function chooseSegmentTool(finish: (kind: SegmentKind) => void) {
  return tool(
    'choose_segment',
    'Commit to the next segment kind. Call exactly once.',
    { kind: z.enum(['talk', 'music']).describe('what plays next') },
    async (args) => {
      finish(args.kind)
      return { content: [{ type: 'text' as const, text: JSON.stringify({ ok: true }) }] }
    },
  )
}

export type LiveCadenceDeps = {
  // The live mix knobs (spec 12 §3.2): read at every decision point, so the
  // pane's gear lands at the next boundary with no rebuild.
  settings: () => { cadenceMode: 'every_n' | 'random' | 'brain'; musicEveryN: number }
  brain?: Harness
  model?: string
}

// The settings-aware policy: stateless modes are constructed per decision
// (EveryNCadence is two fields), the brain policy — which holds a harness —
// is built once on first use and reused. Without a harness, brain mode falls
// back to every_n at the live depth rather than failing the boundary.
export class LiveCadence implements CadencePolicy {
  private deps: LiveCadenceDeps
  private random = new RandomCadence()
  private brainPolicy: BrainCadence | null = null

  constructor(deps: LiveCadenceDeps) {
    this.deps = deps
  }

  async nextKind(state: CadenceState): Promise<SegmentKind> {
    const { cadenceMode, musicEveryN } = this.deps.settings()
    switch (cadenceMode) {
      case 'every_n':
        return new EveryNCadence(musicEveryN).nextKind(state)
      case 'random':
        return this.random.nextKind(state)
      case 'brain': {
        if (this.deps.brain === undefined) return new EveryNCadence(musicEveryN).nextKind(state)
        this.brainPolicy ??= new BrainCadence({
          brain: this.deps.brain,
          model: this.deps.model ?? '',
          fallback: new EveryNCadence(musicEveryN),
        })
        return this.brainPolicy.nextKind(state)
      }
    }
  }
}

export function buildCadence(
  mode: 'every_n' | 'random' | 'brain',
  { everyN, brain, model = '' }: { everyN: number; brain?: Harness; model?: string },
): CadencePolicy {
  switch (mode) {
    case 'every_n':
      return new EveryNCadence(everyN)
    case 'random':
      return new RandomCadence()
    case 'brain':
      if (brain === undefined) throw new Error('brain cadence requires a harnessed brain')
      return new BrainCadence({ brain, model, fallback: new EveryNCadence(everyN) })
  }
}
